File: app/src/main/java/ru/ls/donkitchen/base/BaseFragment.java
package ru.ls.donkitchen.base;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.trello.rxlifecycle.components.support.RxFragment;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import ru.ls.donkitchen.R;
import ru.ls.donkitchen.activity.base.BaseActivity;
import ru.ls.donkitchen.annotation.ConfigPrefs;

/**
 * Базовый фрагмент
 *
 * @author Lord (Kuleshov M.V.)
 * @since 01.04.2015
 */
public abstract class BaseFragment extends RxFragment {
	@Nullable
	@Bind(R.id.toolbar)
	protected Toolbar toolbar;

	@Inject
	@ConfigPrefs
	SharedPreferences prefs;

	private FragmentSubComponent component;

	public FragmentSubComponent getComponent() {
		return component;
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(getLayoutRes(), container, false);
	}

	@Override
	public void onActivityCreated(@Nullable Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		if (getActivity() instanceof BaseActivity) {
			BaseActivity activity = (BaseActivity) getActivity();
			this.component = activity.getComponent().plus(new FragmentModule(this));

			inject();
		}

		if (toolbar != null) {
			toolbar.setTitle(getActivity().getTitle());
		}

		loadData();
	}

	/**
	 *
	 *
	 * @return
	 */
	protected abstract @LayoutRes int getLayoutRes();

	protected abstract void inject();

	@Override
	public void onViewCreated(View v, Bundle savedInstanceState) {
		super.onViewCreated(v, savedInstanceState);

		initControls(v);
		onRestoreInstanceState(savedInstanceState);
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();

		ButterKnife.unbind(this);
	}

	protected void onRestoreInstanceState(Bundle savedInstanceState) {

	}

	protected void initControls(View v) {
		ButterKnife.bind(this, v);
	}

	protected void loadData() {

	}

	/**
	 * Получение настроек пользователя
	 * @return
	 */
	protected SharedPreferences getPrefs() {
		return prefs;
	}

//	/**
//	 * Получает токен авторизации
//	 *
//	 * @return
//	 */
//	protected String getAuthToken() {
//		return getPrefs().getString(BuildConfig.SP_TOKEN, "");
//	}
}

File: app/src/main/kotlin/ru/ls/donkitchen/fragment/base/BaseFragment.kt
package ru.ls.donkitchen.fragment.base

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.arellomobile.mvp.MvpAppCompatFragment
import kotlinx.android.synthetic.main.toolbar.*
import ru.ls.donkitchen.activity.base.BaseActivity
import ru.ls.donkitchen.activity.base.BaseNoActionBarActivity
import ru.ls.donkitchen.nav.ActivityNavigator
import ru.terrakok.cicerone.NavigatorHolder
import ru.terrakok.cicerone.Router
import javax.inject.Inject

/**
 *
 * @author Lord (Kuleshov M.V.)
 * @since 11.01.16
 */
abstract class BaseFragment: MvpAppCompatFragment() {
    @Inject lateinit var router: Router
    @Inject lateinit var navigatorHolder: NavigatorHolder

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(getLayoutRes(), container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        activity.let { it ->
            if (it is BaseActivity) {
                it.component().inject(this)
            } else if (it is BaseNoActionBarActivity) {
                it.component().inject(this)
            }
        }

        if (toolbar != null) {
            toolbar.title = activity.title
        }

        loadData()
    }

    /**
     * @return
     */
    @LayoutRes protected abstract fun getLayoutRes(): Int

    override fun onPause() {
        navigatorHolder.removeNavigator()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        navigatorHolder.setNavigator(ActivityNavigator(activity))
    }

    override fun onViewCreated(v: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(v, savedInstanceState)

        initControls(v)
        onRestoreInstanceState(savedInstanceState)
    }

    protected open fun onRestoreInstanceState(savedInstanceState: Bundle?) {

    }

    protected open fun initControls(v: View?) {

    }

    protected open fun loadData() {

    }

    protected fun releaseDatabaseHelper() {
//        OpenHelperManager.releaseHelper()
    }
}
